src/compiler/core.ts

```diff
@@ -1285,12 +1285,12 @@ namespace ts {
     export function createFileDiagnostic(file: SourceFile, start: number, length: number, message: DiagnosticMessage): Diagnostic {
         const end = start + length;
 
-        Debug.assert(start >= 0, "start must be non-negative, is " + start);
-        Debug.assert(length >= 0, "length must be non-negative, is " + length);
+        Debug.assert(start >= 0, "start must be non-negative", () => `is ${start}`);

```

Comment:
Adding callbacks is arguably more expensive, as you have to create an environment record allocation each time.

It might make more sense to add something like `Debug.assertNonNegative(start, "start")` or similar.